src/hotspot/share/opto/superword.hpp

```diff
@@ -488,7 +488,94 @@ class SuperWord : public ResourceObj {
+  private:
+    const uint _split_size;
+    const bool _reject;
+    const char* _message;
+
+    SplitTask(const uint split_size, const bool reject, const char* message) :
+        _split_size(split_size), _reject(reject), _message(message)
+    {
+      assert(!reject || split_size == 0, "only non-rejections actually split packs");
+      assert(message != nullptr, "must have message");
+    }
+
+  public:
+    SplitTask(const uint split_size, const char* message) :
+        SplitTask(split_size, false, message) {};
+
+    static SplitTask make_no_split() { return SplitTask(0, "no split"); }
+    static SplitTask make_reject(const char* message) { return SplitTask(0, true, message); }
+    const char* message() const { return _message; }
+    bool is_reject() const { return _reject; }
+
+    int split_size() const {
+      assert(!is_reject(), "only non-rejections have a split size");
+      return _split_size;
+    }
+  };
+
+  // After split_pack, we either have:
+  //    changed    first_pack    second_pack
+  // 1) false      old_pack      nullptr       -> old_pack not modified
+  // 2) true       nullptr       nullptr       -> old_pack rejected
+  // 3) true       old_pack      nullptr       -> old_pack modified (nodes removed)
+  // 4) true       pack1         pack2         -> old_pack split into two packs
+  class SplitStatus {
+  private:
+    Node_List* _first_pack;
+    Node_List* _second_pack;
+    bool _changed;
+
+    SplitStatus(Node_List* first_pack, Node_List* second_pack, bool changed) :
+      _first_pack(first_pack), _second_pack(second_pack), _changed(changed)
+    {
+      assert(_second_pack == nullptr || _changed, "second pack implies is_changed");
+      assert(_second_pack == nullptr || _first_pack != nullptr, "second pack implies first pack");
+      assert(second_pack == nullptr || first_pack != second_pack, "cannot have the same pack twice");
+    }
+
+  public:
+    static SplitStatus make_no_change(Node_List* old_pack) {
+      assert(old_pack != nullptr, "must have 1 pack");
+      return SplitStatus(old_pack, nullptr, false);
+    }
+
+    static SplitStatus make_one_pack(Node_List* first_pack) {
+      assert(first_pack != nullptr, "must have 1 pack");
+      return SplitStatus(first_pack, nullptr, true);
+    }
+
+    static SplitStatus make_two_packs(Node_List* first_pack, Node_List* second_pack) {
+      assert(first_pack != nullptr && second_pack != nullptr, "must have 2 packs");
+      return SplitStatus(first_pack, second_pack, true);
+    }
+
+    static SplitStatus make_reject() {
+      return SplitStatus(nullptr, nullptr, true);
+    }
+
+    bool is_changed() const { return _changed; }
+    Node_List* first_pack() const { return _first_pack; }
+    Node_List* second_pack() const { return _second_pack; }
+  };
+
+  SplitStatus split_pack(const char* split_name, Node_List* pack, SplitTask task);
+  template <typename SplitStrategy>
+  void split_packs(const char* split_name, SplitStrategy strategy);
```

Comment:
Note: I plan to move all of the methods for splitting and filtering to the `PackSet` class in a future refactoring.